File: frontend/js/views/break_premium.js
// Meal & rest break premium generator — one hour of pay per missed break day,
// via /calc/break-premium.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderBreakPremium(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.breakprem.h1.title">// BREAK PREMIUM PAY</span></h1>
        <p class="muted small" data-i18n="view.breakprem.hint.intro">
            Under California Labor Code §226.7, an employer that fails to provide a compliant meal or rest
            break owes the employee one additional hour of pay at the regular rate for each workday the break
            is missed — a separate premium for meal and for rest, so up to two hours per day. It computes the
            meal premium, the rest premium, and the total. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.breakprem.h2.inputs">Premium inputs</h2>
            <form id="breakprem-form" class="inline-form">
                <label><span data-i18n="view.breakprem.label.state">State</span>
                    <input type="text" name="state" value="California" required></label>
                <label><span data-i18n="view.breakprem.label.employer">Employer</span>
                    <input type="text" name="employer_name" value=""></label>
                <label><span data-i18n="view.breakprem.label.employee">Employee</span>
                    <input type="text" name="employee_name" value=""></label>
                <label><span data-i18n="view.breakprem.label.rate">Regular rate ($/hr)</span>
                    <input type="number" step="0.01" min="0" name="regular_rate_usd" value="20" required></label>
                <label><span data-i18n="view.breakprem.label.meal">Meal-break violation days</span>
                    <input type="number" step="1" min="0" name="meal_violation_days" value="10"></label>
                <label><span data-i18n="view.breakprem.label.rest">Rest-break violation days</span>
                    <input type="number" step="1" min="0" name="rest_violation_days" value="5"></label>
                <label><span data-i18n="view.breakprem.label.period">Period</span>
                    <input type="text" name="period_label" value="Q2 2026"></label>
                <label><span data-i18n="view.breakprem.label.date">Statement date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.breakprem.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.breakprem.ph.statute'))}"></label>
            </form>
        </div>
        <div id="breakprem-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#breakprem-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            employer_name: (fd.get('employer_name') || '').trim(),
            employee_name: (fd.get('employee_name') || '').trim(),
            regular_rate_usd: Number(fd.get('regular_rate_usd')) || 0,
            meal_violation_days: Number(fd.get('meal_violation_days')) || 0,
            rest_violation_days: Number(fd.get('rest_violation_days')) || 0,
            period_label: (fd.get('period_label') || '').trim(),
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcBreakPremium(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.breakprem.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#breakprem-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.breakprem.card.total">Total premium</div>
                    <div class="value">${money(doc.total_premium_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.breakprem.card.meal">Meal premium</div>
                    <div class="value">${money(doc.meal_premium_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.breakprem.card.rest">Rest premium</div>
                    <div class="value">${money(doc.rest_premium_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.breakprem.card.hours">Premium hours</div>
                    <div class="value">${doc.total_premium_hours}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="breakprem-copy" type="button" data-i18n="view.breakprem.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="breakprem-download" type="button" data-i18n="view.breakprem.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#breakprem-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.breakprem.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.breakprem.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#breakprem-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'break-premium.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/break_premium.rs
//! Meal & rest break premium pay — under California Labor Code §226.7, an employer
//! that fails to provide a compliant meal or rest break owes the employee one
//! additional hour of pay at the regular rate for each workday the break is missed
//! (a separate premium for meal and for rest, so up to two hours per day). This
//! computes the meal premium, the rest premium, and the total. No existing
//! generator computes per-violation-day premium pay. Drafting aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct BreakPremiumInput {
    pub employer_name: String,
    pub employee_name: String,
    /// Employee's regular rate of pay, per hour.
    pub regular_rate_usd: f64,
    /// Workdays on which a compliant meal break was not provided.
    #[serde(default)]
    pub meal_violation_days: u32,
    /// Workdays on which a compliant rest break was not provided.
    #[serde(default)]
    pub rest_violation_days: u32,
    pub period_label: String,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct BreakPremium {
    pub title: String,
    /// One hour of pay × meal violation days.
    pub meal_premium_usd: f64,
    /// One hour of pay × rest violation days.
    pub rest_premium_usd: f64,
    /// Total premium hours owed (one per violation).
    pub total_premium_hours: u32,
    /// Meal premium + rest premium.
    pub total_premium_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &BreakPremiumInput) -> BreakPremium {
    let meal = cents(i.regular_rate_usd * i.meal_violation_days as f64);
    let rest = cents(i.regular_rate_usd * i.rest_violation_days as f64);
    let total = cents(meal + rest);
    let hours = i.meal_violation_days + i.rest_violation_days;

    let calc_body = format!(
        "At a regular rate of {}, each missed break is one hour of premium pay. {} meal-break violation day(s) = {}; {} rest-break violation day(s) = {}. Total premium: {} hour(s) of pay, or {}.",
        money(i.regular_rate_usd),
        i.meal_violation_days,
        money(meal),
        i.rest_violation_days,
        money(rest),
        hours,
        money(total)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("Break premiums are owed under the wage laws of the State of {}.", i.state)
    } else {
        format!("Break premiums are owed under the wage laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Employer: {}\nEmployee: {}\nPeriod: {}\nStatement date: {}",
                i.employer_name, i.employee_name, i.period_label, i.date
            ),
        },
        DocClause {
            heading: "1. Premium Pay".into(),
            body: "For each workday a compliant meal or rest break was not provided, the employee is owed one additional hour of pay at the regular rate — separately for meal and for rest breaks.".into(),
        },
        DocClause { heading: "2. Calculation".into(), body: calc_body },
        DocClause {
            heading: "3. Payment".into(),
            body: format!(
                "The Employer owes the Employee {} in break-premium pay for the period, payable with the next wage payment.",
                money(total)
            ),
        },
        DocClause { heading: "4. Governing Law".into(), body: governing },
        DocClause {
            heading: "Certification".into(),
            body: format!(
                "Employer: ____________________  Date: __________\n{}",
                i.employer_name
            ),
        },
    ];

    BreakPremium {
        title: "Meal & Rest Break Premium Statement".into(),
        meal_premium_usd: meal,
        rest_premium_usd: rest,
        total_premium_hours: hours,
        total_premium_usd: total,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> BreakPremiumInput {
        BreakPremiumInput {
            employer_name: "Acme Co".into(),
            employee_name: "Pat Worker".into(),
            regular_rate_usd: 20.0,
            meal_violation_days: 10,
            rest_violation_days: 5,
            period_label: "Q2 2026".into(),
            date: "2026-07-01".into(),
            state: "California".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn meal_and_rest_premiums() {
        let d = generate(&base());
        assert!(close(d.meal_premium_usd, 200.0));
        assert!(close(d.rest_premium_usd, 100.0));
        assert_eq!(d.total_premium_hours, 15);
        assert!(close(d.total_premium_usd, 300.0));
    }

    #[test]
    fn no_violations_zero() {
        let d = generate(&BreakPremiumInput { meal_violation_days: 0, rest_violation_days: 0, ..base() });
        assert!(close(d.total_premium_usd, 0.0));
        assert_eq!(d.total_premium_hours, 0);
    }

    #[test]
    fn fractional_rate() {
        let d = generate(&BreakPremiumInput { regular_rate_usd: 33.33, meal_violation_days: 8, rest_violation_days: 8, ..base() });
        assert!(close(d.meal_premium_usd, 266.64));
        assert!(close(d.total_premium_usd, 533.28));
    }

    #[test]
    fn meal_only() {
        let d = generate(&BreakPremiumInput { rest_violation_days: 0, ..base() });
        assert!(close(d.rest_premium_usd, 0.0));
        assert!(close(d.total_premium_usd, 200.0));
        assert_eq!(d.total_premium_hours, 10);
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&BreakPremiumInput { statute_citation: "Cal. Lab. Code § 226.7".into(), ..base() });
        assert_eq!(d.statutory_citation, "Cal. Lab. Code § 226.7");
        assert!(d.clauses.iter().any(|c| c.body.contains("Cal. Lab. Code § 226.7")));
    }
}
